include/deal.II/base/exceptions.h

```diff
@@ -1519,35 +1519,39 @@ namespace deal_II_exceptions
 } /*namespace deal_II_exceptions*/
 
 
-#if defined(__clang__)
-#  define DEAL_II_ASSUME(expr) __builtin_assume(static_cast<bool>(expr))
-#elif defined(__GNUC__) && !defined(__ICC)
-#  if __GNUC__ >= 13
-#    define DEAL_II_ASSUME(expr)                                         \
-      do                                                                 \
-        {                                                                \
-          _Pragma("GCC diagnostic push")                                 \
-            _Pragma("GCC diagnostic ignored \"-Wimplicit-fallthrough\"") \
-              [[assume(expr)]];                                          \
-          _Pragma("GCC diagnostic pop")                                  \
-        }                                                                \
-      while (false)
-#  else
+#ifdef DEAL_II_HAVE_CXX23
+#  define define DEAL_II_ASSUME(expr) [[assume(expr)]]
+#else
+#  if defined(__clang__)
+#    define DEAL_II_ASSUME(expr) __builtin_assume(static_cast<bool>(expr))
+#  elif defined(__GNUC__) && !defined(__ICC)
+#    if __GNUC__ >= 13
+#      define DEAL_II_ASSUME(expr)                                         \
+        do                                                                 \
+          {                                                                \
+            _Pragma("GCC diagnostic push")                                 \
+              _Pragma("GCC diagnostic ignored \"-Wimplicit-fallthrough\"") \
+                [[assume(expr)]];                                          \
+            _Pragma("GCC diagnostic pop")                                  \
+          }                                                                \
+        while (false)
+#    else
 /* no way with GCC to express this without evaluating 'expr' */
+#      define DEAL_II_ASSUME(expr) \
+        do                         \
+          {                        \
+          }                        \
+        while (false)
+#    endif
+#  elif defined(_MSC_VER) || defined(__ICC)
+#    define DEAL_II_ASSUME(expr) __assume(expr);
+#  else
 #    define DEAL_II_ASSUME(expr) \
       do                         \
         {                        \
         }                        \
       while (false)
 #  endif
```

Comment:
We still need to disable the warnings for `gcc-13` so I would rather modify this last case
```suggestion
#  elif defined(DEAL_II_HAVE_CXX23)
#    define DEAL_II_ASSUME(expr) [[assume(expr)]]
#  else
#    define DEAL_II_ASSUME(expr) \
      do                         \
        {                        \
        }                        \
     while (false)
```

Reply:
I didn't actually change anything about the existing cases, github just has a hard time showing it as such. Take a look here: https://github.com/dealii/dealii/pull/16457/files?w=1

Reply:
Yes, I know. My point was that we might already say that g++-13 supports C++23 to some extent and we would then not suppress the warnings with `switch` constructs.

Reply:
I must admit that I'm confused. I put the `ifdef DEAL_II_HAVE_CXX23` case at the very top, because then we have a standardized way of saying what we want to say. All the rest is in the `#else` branch. Your suggested addition of
```
#  elif defined(DEAL_II_HAVE_CXX23)
#    define DEAL_II_ASSUME(expr) [[assume(expr)]]
#  else
```
further down is a dead branch that cannot be reached.

Reply:
My only concern is that we run into the new branch instead of the GCC one if we decide that gcc-13 provides sufficient C++23 support. In that case, we would get a lot of warnings that I explicitly suppressed. We should also test for this feature as we usually do when setting DEAL_II_HAVE_CXX23 anyway.

Reply:
I *think* what you're saying is that you want us to use the GCC-specific way even if `[[assume]]` is available? Why would we do this? If there's a standard way to do something, we should use it.